crypto/fipsmodule/rand/new_rand.c

```diff
@@ -0,0 +1,290 @@
+    &first_pred_resistance_len);
+
+  // If caller input user-controlled prediction resistance, use it.
+  if (use_user_pred_resistance == RAND_USE_USER_PRED_RESISTANCE) {
+    for (size_t i = 0; i < RAND_PRED_RESISTANCE_LEN; i++) {
+      pred_resistance[i] ^= user_pred_resistance[i];
+    }
+    first_pred_resistance_len = RAND_PRED_RESISTANCE_LEN;
+  }
+
+  assert(first_pred_resistance_len == 0 ||
+         first_pred_resistance_len == RAND_PRED_RESISTANCE_LEN);
+
+  // Iterate CTR-DRBG generate until we generated |out_len| bytes of randomness.
+  while (out_len > 0) {
+    size_t todo = out_len;
+    if (todo > CTR_DRBG_MAX_GENERATE_LENGTH) {
+      todo = CTR_DRBG_MAX_GENERATE_LENGTH;
+    }
+
+    // Each reseed interval can generate up to
+    // |CTR_DRBG_MAX_GENERATE_LENGTH*2^{kCtrDrbgReseedInterval}| bytes.
+    // Determining the time(s) to reseed prior to entering the CTR-DRBG generate
+    // loop is a doable strategy. But tracking reseed times adds unnecessary
+    // complexity. Instead our strategy is optimizing for simplicity.
+    // |out_len < CTR_DRBG_MAX_GENERATE_LENGTH| will be the majority case
+    // (by far) and requires a single check in either strategy.
+    // Note if we just reseeded through |rand_ctr_drbg_reseed()|, we won't
+    // reseed again here.
+    if( state->generate_calls_since_seed + 1 >= kCtrDrbgReseedInterval) {
+      rand_ctr_drbg_reseed(state);
+    }
+
+    if (!CTR_DRBG_generate(&state->drbg, out, todo, pred_resistance,
+          first_pred_resistance_len)) {
+      abort();
+    }
+
+    out += todo;
+    out_len -= todo;
+    state->generate_calls_since_seed++;
+    first_pred_resistance_len = 0;
+  }
+
+  OPENSSL_cleanse(pred_resistance, RAND_PRED_RESISTANCE_LEN);
+
+  if (rand_ensure_valid_state() != 1) {
+    abort();
+  }
+}
+
+static void RAND_bytes_private(uint8_t *out, size_t out_len,
+  const uint8_t user_pred_resistance[RAND_PRED_RESISTANCE_LEN],
+  int use_user_pred_resistance) {
+
+  if (out_len == 0) {
+    return;
+  }
+
+  struct rand_thread_local_state *state =
+      CRYPTO_get_thread_local(OPENSSL_THREAD_LOCAL_PRIVATE_RAND);
+
+  if (state == NULL) {
+    state = OPENSSL_zalloc(sizeof(struct rand_thread_local_state));
+    if (state == NULL ||
+        CRYPTO_set_thread_local(OPENSSL_THREAD_LOCAL_PRIVATE_RAND, state,
+                                   rand_thread_local_state_free) != 1) {
+      abort();
+    }
+
+    rand_state_initialize(state);
+  }
+
+  RAND_bytes_core(state, out, out_len, user_pred_resistance,
+    use_user_pred_resistance);
+}
+
+// TOOD
+// Retire and replace call sites with RAND_bytes_with_user_prediction_resistance
+int NR_PREFIX(RAND_bytes_with_additional_data)(uint8_t *out, size_t out_len,
+  const uint8_t user_pred_resistance[RAND_PRED_RESISTANCE_LEN]) {
+  
+  RAND_bytes_private(out, out_len, user_pred_resistance,
+    RAND_USE_USER_PRED_RESISTANCE);
+  return 1;
+}
+
+int RAND_bytes_with_user_prediction_resistance(uint8_t *out, size_t out_len,
```

Comment:
Isn't this the same as RAND_bytes_with_additional_data?

Reply:
Yes, correct. But I want to correct the name.

The `additional_data` was not really used for "additional data" in the sense of SP800-90. It was used as prediction resistance.

Reply:
Are you planning on making this a public API? Will that be even more confusing for future customers? They'll have to pick from public/private random sources, and prediction resistance or additional data. How should they figure out what to use?